DI: Consider capture of `let`s to be a read-only use.

jckarter opened the pull request:
SILGen emits capture of an address-only `let` property as @inout_aliasable, but Sema enforces that the closure is not really allowed to modify the property, so DI can consider the capture to be a read. Fixes rdar://problem/40828667.

A more principled solution would be to treat `let` captures as @in_guaranteed (see #17047), but that unfortunately leads to breakage elsewhere in the optimizer.

jckarter commented:
@swift-ci Please smoke test

gottesmm commented on lib/SILOptimizer/Mandatory/DIMemoryUseCollectorOwnership.cpp:
```diff
@@ -877,9 +877,30 @@ void ElementUseCollector::collectUses(SILValue Pointer, unsigned BaseEltNo) {
+        // a nonmutating use.
+        bool isLet = true;
+        
+        for (unsigned i = 0; i < TheMemory.NumElements; ++i) {
```
Can you use none_of here instead? i.e.:

```
if (llvm::none_of(indices(TheMemory.NumElements),
                            [&](unsigned i) { return !TheMemory.isElementLetProperty(i) })) {
  addElementUses(...);
  continue;
}
```

jckarter replied:
Thanks for the suggestion, but I don't find that easier to read.

gottesmm replied:
I am not going to push on this... but I just wanted to point out that I was just bringing up this section from the LLVM style guide:

https://www.llvm.org/docs/CodingStandards.html#turn-predicate-loops-into-predicate-functions

jckarter replied:
Even that doesn't recommend using `none_of`, though, but splitting the loop out into a separate function,

gottesmm commented on lib/SILOptimizer/Mandatory/PMOMemoryUseCollector.cpp:
```diff
@@ -555,8 +555,28 @@ bool ElementUseCollector::collectUses(SILValue Pointer, unsigned BaseEltNo) {
+        // a nonmutating use.
+        bool isLet = true;
+        
+        for (unsigned i = 0; i < TheMemory.NumElements; ++i) {
```
Same here.

gottesmm replied:
Also, please add a test case for PMO.

jckarter replied:
What is PMO?

gottesmm replied:
PMO == Predictable Memory Opts. Predictable Mem Opts and DI now have two different use collector impls. So if you only care about DI, just change the Ownership one. I am waiting for you to finish this to rename the Ownership version to not have the Ownership suffix.

If your thing doesn't care about predictable memory opts, then I wouldn't change this.

jckarter replied:
You may want to wait to change the name until we wrap up the 4.2 branch, since that will make cherry-picking changes over more difficult and error-prone. Does PMO affect diagnostics?

jckarter replied:
Furthermore, does the formerly-DIMemoryUseCollector still have a role in DI on the 4.2 branch?

gottesmm replied:
No PMO does not. It is only used for predictable memory operations. The idea was to split the two passes so that they could evolve independently. They had a bunch of non-shared code that made it hard to reason about.

I think that git is good enough to handle moves... but I can wait.

jckarter replied:
My concern with the move here is that you'd be replacing a file with a different-but-similar-enough file that Git might mistake it for an in-place modification rather than a sequence of renames.

jckarter commented:
@swift-ci Please smoke test

jckarter commented:
@swift-ci Please test source compatibility

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test source compatibility

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test source compatibility

jckarter commented:
@swift-ci Please test source compatibility

jckarter committed:
Sema: Pass the useDC down to getTypeOfReference.

This fixes a subtle issue where, during single-expression closure type inference, we would ask for the settability of local variables from the outer function's context, leading us to mistakenly consider them mutable inside single-expression closure contexts. DI would catch some but not all violations of the expected semantics here.

jckarter committed:
DI: Consider captures of let properties to be read-only uses.

Sema enforces that closures can't modify the let properties of their captured contexts, but we consider the capture argument to be @inout_aliasable at the SIL level. Add an exception to our normal handling so that this use is considered read-only. Fixes rdar://problem/40828667.

A more principled solution would be to treat let captures as @in_guaranteed (see #17047), but that unfortunately leads to breakage elsewhere in the optimizer.

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test source compatibility

jckarter commented:
@swift-ci Please test source compatibility

slavapestov commented on test/SILOptimizer/definite_init_diagnostics.swift:
```diff
@@ -83,11 +83,6 @@ func test2() {
   }
   b4 = 7
   
-  let b5: Any
```
Why did you remove this test case?

jckarter replied:
The change to Sema makes the DI test redundant, and furthermore, since the error here gets diagnosed in Sema, it prevents any other DI tests in this file from being reached.

jckarter commented:
@swift-ci Please test OS X
